test/gurka/test_matrix_time_dependent.cc

```diff
@@ -0,0 +1,227 @@
+                                       {"E", "H"}, "auto", {}, {}, &res);
+
+  rapidjson::Document res_doc;
+  res_doc.Parse(res.c_str());
+
+  // we expect to take the motorways, residential path is 1.6f
+  check_dist(res_doc, {0.0f, 2.0f, 2.0f, 0.0f});
+}
+
+TEST_F(MatrixTest, MatrixWithLiveTraffic) {
+  allowed_map.config.put("mjolnir.traffic_extract", "test/data/matrix_traffic_allowed/traffic.tar");
+
+  test::build_live_traffic_data(allowed_map.config);
+  std::shared_ptr<valhalla::baldr::GraphReader> reader =
+      test::make_clean_graphreader(allowed_map.config.get_child("mjolnir"));
+
+  test::LiveTrafficCustomize edges_with_traffic = [](baldr::GraphReader& reader,
+                                                     baldr::TrafficTile& tile, uint32_t index,
+                                                     baldr::TrafficSpeed* current) -> void {
+    // update speeds on primary road
+    update_traffic_on_edges(reader, tile, index, current, "FA", allowed_map, 5);
+    update_traffic_on_edges(reader, tile, index, current, "AB", allowed_map, 5);
+    update_traffic_on_edges(reader, tile, index, current, "BG", allowed_map, 5);
+  };
+  test::customize_live_traffic_data(allowed_map.config, edges_with_traffic);
+
+  const std::vector<float> exp_dists = {0.0f, 1.6f, 1.6f, 0.0f};
+  std::unordered_map<std::string, std::string> options = {{"/date_time/type", "0"}};
+
+  // forward tree
+  std::string res;
+  auto result = gurka::do_action(Options::sources_to_targets, allowed_map, {"E", "H"}, {"E", "H"},
+                                 "auto", options, reader, &res);
+  rapidjson::Document res_doc;
+  res_doc.Parse(res.c_str());
+  // we expect to take the shorter lower path on residential
+  check_dist(res_doc, exp_dists);
+  ASSERT_EQ(result.info().warnings().size(), 0);
+
+  // forward tree, date_time on the locations
+  options = {{"/sources/0/date_time", "current"},
+             {"/sources/1/date_time", "2016-07-03T08:06"},
+             {"/costing_options/auto/speed_types/0", "current"}};
+  res.erase();
+  result = gurka::do_action(Options::sources_to_targets, allowed_map, {"E", "H"}, {"E", "H"}, "auto",
+                            options, nullptr, &res);
+  res_doc.Parse(res.c_str());
+  // the second origin can't respect time (no historical data)
+  check_dist(res_doc, {0.0f, 1.6f, 2.0f, 0.0f});
+  ASSERT_EQ(result.info().warnings().size(), 0);
+
+  // reverse tree, no time_info
+  res.erase();
+  result = gurka::do_action(Options::sources_to_targets, allowed_map, {"E", "H"}, {"H"}, "auto",
+                            options, reader, &res);
+  res_doc.Parse(res.c_str());
+  // without time, we'd expect to take the longer, faster route
+  // TODO: be aware that it's actually 2.0f, but CostMatrix has a bug here
+  check_dist(res_doc, {1.6f, 0.0f});
+  ASSERT_EQ(result.info().warnings().size(), 1);
+  ASSERT_EQ(result.info().warnings().Get(0).code(), 401);
```

Comment:
here we have two origins with date_time but perform a reverse search, so we can't respect it. in that case it'll fall back to CostMatrix which should actually take the faster and longer route, but it doesn't. thankfully I've seen it in other tests already that there's a bug, e.g. https://github.com/valhalla/valhalla/pull/3792#issuecomment-1279963888. ref to issue: https://github.com/valhalla/valhalla/issues/3803